server/remote_cache/cachetools/cachetools.go

```diff
@@ -269,6 +276,232 @@ func UploadProtoToCAS(ctx context.Context, cache interfaces.Cache, instanceName
 	return uploadProtoToCache(ctx, cas, instanceName, in)
 }
 
+// BatchCASUploader uploads many files to CAS concurrently, batching small
+// uploads together and falling back to bytestream uploads for large files.
+type BatchCASUploader struct {
+	ctx              context.Context
+	eg               *errgroup.Group
+	cache            interfaces.Cache
+	bsClient         bspb.ByteStreamClient
+	casClient        repb.ContentAddressableStorageClient
+	req              *repb.BatchUpdateBlobsRequest
+	currentBatchSize int64
+	instanceName     string
+	uploads          map[digest.Key]struct{}
+}
+
+func NewBatchCASUploader(ctx context.Context, env environment.Env, instanceName string) (*BatchCASUploader, error) {
+	bsClient := env.GetByteStreamClient()
+	if bsClient == nil {
+		return nil, status.InvalidArgumentError("Missing bytestream client")
+	}
+	casClient := env.GetContentAddressableStorageClient()
+	if casClient == nil {
+		return nil, status.InvalidArgumentError("Missing CAS client")
+	}
+	cache := env.GetCache()
+	if cache == nil {
+		return nil, status.InvalidArgumentError("Missing cache")
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+	return &BatchCASUploader{
+		ctx:              ctx,
+		eg:               eg,
+		cache:            cache,
+		bsClient:         bsClient,
+		casClient:        casClient,
+		req:              &repb.BatchUpdateBlobsRequest{InstanceName: instanceName},
+		currentBatchSize: 0,
+		instanceName:     instanceName,
+		uploads:          make(map[digest.Key]struct{}),
+	}, nil
+}
+
+// Upload adds the given content to the current batch or begins a streaming
+// upload if it exceeds the maximum batch size. It closes r when it is no
+// longer needed.
+func (ul *BatchCASUploader) Upload(d *repb.Digest, r io.ReadSeekCloser) error {
+	// De-dupe uploads by digest.
+	dk := digest.NewKey(d)
+	if _, ok := ul.uploads[dk]; ok {
+		return r.Close()
+	}
+	ul.uploads[dk] = struct{}{}
+
+	r.Seek(0, 0)
+
+	if d.GetSizeBytes() > gRPCMaxSize {
+		ul.eg.Go(func() error {
+			_, err := UploadFromReader(ul.ctx, ul.bsClient, digest.NewInstanceNameDigest(d, ul.instanceName), r)
+			if err != nil {
+				return err
+			}
+			return r.Close()
```

Comment:
maybe we should defer r.Close() as the first line of the closure body so it isn't leaked if UploadFromReader returns some other error?